Add a type to the v1 model for the updated ClusterConfig

jan-g opened the pull request:
This is a placeholder commit to get feedback before we go too far down that path - we'll add the CRD
definition once we get agreement on it.

jan-g commented:
This is mostly to invite comment about the proposed attribute for cluster config (we'll do something similar for node configuration rather than splitting the AdditionalConfiguration up according to prefix).

I think this needs to be merged with the current configuration - but because we want late binding of external secrets, we'll have to teach the configuration subcommand to pre-process a `bootstrap.yaml.in` to locate and expand these references to generate a `.bootstrap.yaml`, as well as managing the expansion in the operator itself for dynamic configuration update.

alenkacz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -192,6 +198,39 @@ type ClusterSpec struct {
+	// If the value is supplied by an external source, coordinates are embedded here.
+	// For non-string target types, the string value fetched from the source will be treated as
+	// a value encoded according to YAML rules.
+	ExternalSecretRef *struct {
```
the original agreement was that this feature will pretty much be hidden, also I think we'll be using just secret name as a generic/platform agnostic pointer to what you want to read.

We can also emit this from the first implementation.

For reference, what I am going to get from controlplane is `${secrets.PASSWORD}`

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -192,6 +198,45 @@ type ClusterSpec struct {
+// given key, the corresponding values here should be string-encoded (according to yaml
+// rules)
+// TODO: this type should be lifted to a shared module rather than duplicated in the `redpanda` CRD definition.
+type ClusterConfiguration struct {
```
Do we expect there to be members other than `Values` here? If not, could this just be `map[string]ClusterConfigValue`?

jan-g replied:
I think probably having an `",omitempty"` annotation on this is appropriate - it'd let us come back and slide in typed fields later (issues with the generation tooling require monomorphic types notwithstanding).

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -166,6 +166,12 @@ type ClusterSpec struct {
 	// By default if Replicas is 3 or more and redpanda.default_topic_partitions is not set
 	// default webhook is setting redpanda.default_topic_partitions to 3.
 	AdditionalConfiguration map[string]string `json:"additionalConfiguration,omitempty"`
+	// Cluster configuration values may also be held here. A `keyName` entry in this
```
Is the distinction between `keyName` and `keyname` intentional / important here?

jan-g replied:
It's not, fixed.

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -192,6 +198,45 @@ type ClusterSpec struct {
+	Values map[string]ClusterConfigValue `json:"values,omitempty"`
+}
+
+type YamlRepresentation string
```
```suggestion
type YAMLRepresentation string
```

nit but go's naming conventions would recommend this.

A comment and some examples would go a long way here too. Though it's possible example may cause more confusion as:
```
Foo
"Foo"
True
true
```

Are all valid

jan-g replied:
I'll look to get at least one entry here with a kuttl test or something going, just to demonstrate that it works e2e.

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -192,6 +198,45 @@ type ClusterSpec struct {
+	// Note: during the construction of a bootstrap template file, references to config objects above will be
+	// rewritten to an EnvVarRef instead, and that variable added to the container environment via the usual k8s
+	// envFrom mechanism. A preferred environment variable name may be specified here.
+	EnvVarRef *string `json:"envVar,omitempty"`
```
I don't think we want to directly expose this, do we? We can rewrite the `{ConfigMap,Secret}KeyRef` as need be ourselves.

jan-g replied:
This is down to me reusing the type, but I'll probably rejig that since the variations show up in different places.

jan-g replied:
Fixed up.

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -166,6 +166,12 @@ type ClusterSpec struct {
+	// this is to permit the migration of those settings to this attribute.
+	// The configuration may contain references to values extracted from k8s ConfigMaps or Secrets;
+	// furthermore, we support the fetching of provider-specific secrets directly.
+	ClusterConfiguration ClusterConfiguration `json:"clusterConfiguration,omitempty"`
```
This won't matter if ClusterConfiguration becomes a map type but `omitempty` doesn't work on structs.

jan-g replied:
That's fixed now.

chrisseto commented on operator/cmd/configurator/configurator.go:
```diff
@@ -57,6 +57,8 @@ const (
 	redpandaRPCPortEnvVar                                = "REDPANDA_RPC_PORT"
 	svcFQDNEnvVar                                        = "SERVICE_FQDN"
 	additionalListenersEnvVar                            = "ADDITIONAL_LISTENERS"
+	bootstrapDestinationEnvVar                           = "BOOTSTRAP_DESTINATION"
```
just my two cents but I think pulling these from CLI args would be a bit nicer for debuggers. We can punt this to a refactor later as the configurator desperately needs something to help read configs from CLI and/or Envvars

jan-g replied:
Agreed. I just followed the existing pattern here which I guess is strongly "12-factor"-influenced, but I'd prefer something better here.

chrisseto commented on operator/pkg/resources/configmap.go:
```diff
@@ -380,6 +385,25 @@ func (r *ConfigMapResource) CreateConfiguration(
+	cfg.BootstrapConfiguration = make(map[string]vectorizedv1alpha1.ClusterConfigValue, len(cfg.ClusterConfiguration))
+	for k, v := range cfg.ClusterConfiguration {
+		// These values are all "concrete" - that is, they're not looked up anywhere.
+		buf, err := json.Marshal(v)
```
Maybe add a note here to acknowledge that this is intentionally JSON and not YAML?

jan-g replied:
Done

chrisseto commented on operator/pkg/resources/configuration/expander.go:
```diff
@@ -0,0 +1,105 @@
+// ExpandForBootstrap passes over the input ClusterConfiguration and returns one with
+// all ConfigMapKeyRef and SecretKeyRef entries turned into environment references,
+// together with a list of env var definitions required to satisfy those.
+func ExpandForBootstrap(cfg map[string]vectorizedv1alpha1.ClusterConfigValue) (map[string]vectorizedv1alpha1.ClusterConfigValue, []corev1.EnvVar, error) {
```
I think having a distinction between the CRD API and the intermediate template/expandable type would be useful. `ConfigMapKeyRef` and `SecretKeyRef` aren't valid in the template expansion case and `EnvVarRef` isn't really valid in the CRD API case.

jan-g replied:
Yeah, done.

chrisseto commented on operator/pkg/resources/configuration/expander.go:
```diff
@@ -0,0 +1,105 @@
+				ExternalSecretRef: v.ExternalSecretRef,
+			}
+		default:
+			return nil, nil, fmt.Errorf("bad cluster config value for %s", k)
```
This error is pretty ambiguous. How about:

```suggestion
			return nil, nil, fmt.Errorf("cluster config %q has no value: %#v", k, v)
```

chrisseto commented on operator/pkg/resources/statefulset.go:
```diff
@@ -111,7 +112,10 @@ type StatefulSetResource struct {
+	nodeConfigMapHashGetter func(context.Context) (string, error)
+	// The configuration we've installed. It's injected so that we can extract any additional
+	// environment variables the configuration requires to expand the bootstrap template.
+	configurationGetter      func(ctx context.Context) (*configuration.GlobalConfiguration, error)
```
I'm not sure what contexts this is used in but could this just be `[]corev1.EnvVar` rather than a closure to limit how much of the configuration "leaks" into the STS generation?

alenkacz reviewed (commented):
Chris made very good comments 👌 I like the direction

jan-g commented:
One last bit to take care of: the older GlobalConfiguration.ClusterConfiguration really isn't enough to determine whether the live admin api should be poked.

Update last-applied-configuration to hold the hash of (the all-values) concretized cluster config, together with the resulting cluster config version.

jan-g commented:
The GlobalConfiguration api is still a mess - it's used as an open structure by a lot of code, and we need (for the moment) to be backwards-compatible with the way config was initially processed.

To follow on from this (but not in this PR):

- ditch drift configuration and just reassert state on a regular basis (this is a no-op if there's no change)
- split the configmap and configuration building out into separate pieces; split node configuration and cluster configuration into distinct notions and stop exposing internal fields
- drop the mixed-mode and classic mode support in favour of bootstrap configuration

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -196,6 +202,43 @@ type ClusterSpec struct {
+// rules)
+type ClusterConfiguration map[string]ClusterConfigCRDValue
+
+type YAMLRepresentation string
```
nit: a comment here could be useful

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -196,6 +202,43 @@ type ClusterSpec struct {
+// throughout. The octet sequence of a Representation will be inserted into a bootstrap template
+// verbatim.
+// TODO: this type should be lifted to a shared module rather than duplicated in the `redpanda` CRD definition.
+type ClusterConfigCRDValue struct {
```
nit: Feels funky to include CRD in the name of a value that's included as part of a CRD. _Hopefully_ the package name and/or some kubebuilder annotations are enough to showcase that this is part of a CRD.
```suggestion
type ClusterConfigValue struct {
```

chrisseto replied:
OOC what is the expected behavior if all provided members are `nil`? If that's an invalid case could you add a TODO to add in CEL validation thereof?

jan-g replied:
Done

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -196,6 +202,43 @@ type ClusterSpec struct {
+	// The string "foo" should be the five octets "\"foo\""
+	// A true value should be the four octets "true".
+	// The number -123456 should be a seven-octet sequence, "-123456".
+	Representation *YAMLRepresentation `json:"repr,omitempty"`
```
nit: I think not having matching field names and JSON tags is going to trip someone up along the way. Mind renaming the field to `Repr` (or `Value`) or making everyone writing out YAML have to fully type out `representation`?

jan-g replied:
I'll use Repr rather than Value (I'd like the attribute name to slap people in the face with the expectation here.)

chrisseto commented on operator/cmd/configurator/bootstrap.go:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	var config []string
+	for k, v := range template {
```
I don't think this will ever matter but do you think we should be concerned that the key ordering could get shuffled around here?

jan-g replied:
I think probably not, the yaml input is an object. If there's some implicit ordering we can fix that up. The result is only visible to the local process.

chrisseto commented on operator/cmd/configurator/bootstrap.go:
```diff
@@ -0,0 +1,36 @@
+		if err != nil {
+			return fmt.Errorf("cannot resolve value %s: %w", k, err)
+		}
+		config = append(config, fmt.Sprintf("%s: %s", k, repr))
```
This hurts me a bit to see but I think it'll be fine for all cases we're interested in. If anything, we could drop all of this into a `map` and implement `MarshalYAML/JSON` for the repr type so we could delegate to `yaml.Marshal`.

jan-g replied:
I had a brief look at this - MarshalYAML returns an interface{} that's then marshalled out, which means if you want it to work then you need the type of the thing to return (which puts us back where we started). There didn't appear to be anything to hand that'd let me supply a []byte.

jan-g replied:
(Yes, this made me wince too.)

chrisseto reviewed (approved):
Nothing blocking, nits are nice to haves and/or can be handled later :shipit:

jan-g commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|release/v2.4.x|[<img src="https://img.shields.io/github/pulls/detail/state/redpanda-data/redpanda-operator/616">](https://github.com/redpanda-data/redpanda-operator/pull/616)|
|✅|release/v2.3.x|[<img src="https://img.shields.io/github/pulls/detail/state/redpanda-data/redpanda-operator/617">](https://github.com/redpanda-data/redpanda-operator/pull/617)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
